Consolidate post flight and non-essential save calls in to a single save action

MrD-RC opened the pull request:
Fixes #3838 

This PR reduces the number of non-essential saves processed by the flight controller. Currently, all calls to `saveConfigAndNotify()` are processed immediately. This means that multiple saves can be processed in the same loop. This PR changes that function to set a flag, to process the save in the next loop. Meaning multiple save calls are now consolidated in to a single save action. For example, continuous servo trim and stats would both call save. Now, there would be a single save for both features. Saving stats and continuous servo trim have been moved over to `saveConfigAndNotify()`. Another function `saveConfig();` has been added for saves where notification is not needed. Going forward, anything that saves on disarm should use `saveConfigAndNotify()`. Other saves, currently using `writeEEPROM()`, that do not reboot after saving; can instead use `saveConfig()` to schedule and consolidate actual saves.

Also, messages has been added to the OSD. A `** SAVING SETTINGS **` messages appears when the save is requested. Then a `** SETTINGS SAVED **` message takes over for 5 seconds; once the save command has completed. These are shown with both standard `SYSTEM MESSAGES` and on the stats screens.

![image](https://user-images.githubusercontent.com/17590174/193416449-c805c881-0ac2-481e-8653-a84b87ec769c.png)

![image](https://user-images.githubusercontent.com/17590174/193416418-f87d4652-c4d8-475e-84c6-0568ff4d6b78.png)

MrD-RC committed:
Reduce post flight and non-essential saves

This PR reduces the number of non-essential saves processed by the flight controller. Currently, all calls to `saveConfigAndNotify()` are processed immediately. This means that multiple saves can be processed in the same loop. This PR changes that function to set a flag, to process the save in the next loop. Meaning multiple save calls are now consolidated in to a single call. For example, continuous servo trim and stats would both call save. Now, there would be a single save for both features. Saving stats and continuous servo trim have been moved over to `saveConfigAndNotify()`.

Also, messages has been added to the OSD. A `** SAVING SETTINGS ** messages appears when the save is requested. Then a `** SETTINGS SAVED **` message takes over for 5 seconds; once the save command has completed. These are shown with both standard `SYSTEM MESSAGES` and on the stats screen.

breadoven commented on src/main/fc/fc_tasks.c:
```diff
@@ -410,6 +412,12 @@ cfTask_t cfTasks[TASK_COUNT] = {
         .desiredPeriod = TASK_PERIOD_HZ(10),              // run every 100 ms, 10Hz
         .staticPriority = TASK_PRIORITY_HIGH,
     },
+    [TASK_SAVE_SETTINGS] = {
```
This seems a bit like overkill for something that only happens once per flight at the point of disarm. Is there not a more efficient way of doing it using a flag set at Arming/Disarming that will achieve the same thing ?

MrD-RC replied:
Sure. But I couldn't see anywhere relevant to system functions that is called at the start of the loop. Any suggestions are welcome. All that needs to happen is the 'processDelayedSave();' is called. 

This is not called on disarm. It is not necessarily only on disarm. For example the CMS also uses saveConfigAndNotify. It could also be expanded, so non-immediate saveEEPROM calls are consolidated in to this feature.

MrD-RC replied:
I've removed the `SAVE_SETTINGS` task and put the call in to `SYSTEM`. This seems like the most suitable option of all the running tasks.

breadoven commented on src/main/fc/config.c:
```diff
@@ -177,8 +181,7 @@ uint32_t getGyroLooptime(void) {
 }
 
-void validateAndFixConfig(void)
-{
```
Why has the format for functions been changed ? Is this some sort of auto format change or deliberate ?

MrD-RC replied:
Just for consistency.

stronnag replied:
The historic INAV standard has been "brace on new line", and is documented as such via https://github.com/iNavFlight/inav/blob/master/docs/development/Development.md#general-principals (and a link to the basefight coding standard).

So the above is only really consistent with  the other inconsistencies in the source file, if we care about such things.

MrD-RC replied:
Interesting. It goes back to multiwii. A bit inconsistent between functions and logic, but fair enough. Them's the rules. 

_Curly brace on same line for everything except function definitions_

I'll make it so.

MrD-RC committed:
Fixed code styling

MrD-RC committed:
Merge branch 'master' into MrD_Minimise-save-calls-and-show-notification

MrD-RC committed:
Added missing function for saveConfig()

This can be used instead of writeEEPROM() if the FC is not rebooted after writing. This will also consolidated multiple saves.

MrD-RC committed:
Merge branch 'MrD_Minimise-save-calls-and-show-notification' of https://github.com/iNavFlight/inav into MrD_Minimise-save-calls-and-show-notification

MrD-RC committed:
Moved processDelayedSave call

- removed `SAVE` task
- moved `processDelayedSave()` call in to `SYSTEM`

MrD-RC committed:
Moved save function call to only when disarmed
